### ISSUE TITLE
[css-images-4] Should `contain-intrinsic-size` affect `object-fit`?

### ISSUE THREAD

**Issue by eeeps** [w3c/csswg-drafts#10116]
[This question came up during `sizes=auto` spec/implementation work](https://github.com/whatwg/html/issues/10079), and I don't know the answer.

Here's a test page: https://codepen.io/eeeps/pen/mdgWNJo?editors=1101

[WebKit thinks that it should](https://o.img.rodeo/image/upload/v1711052026/kspptzsafh9kex0rtnf7.png), [Chromium thinks that it should](https://o.img.rodeo/image/upload/v1711052050/eyr42shpztgsxccigtr1.png), and [Firefox refuses to do any kind of object-fitting on size-contained elements](https://o.img.rodeo/image/upload/v1711052061/zfnvwrz2watd9exuah6i.png).

Notably, @mirisuzanne [thinks that it shouldn't,](https://github.com/whatwg/html/issues/10079#issuecomment-2010705829) because `contain-intrinsic-size` doesn't actually change the natural dimensions of replaced elements (`Image.naturalWidth`/`naturalHeight` are unaffected), but simply tells the layout algorithm to proceed *as if those were the natural dimensions*.

I agree with her, and think UAs should use the actual natural dimensions that you get from `Image.naturalWidth`/`naturalHeight`, and not the `contain-intrinsic-size` dimensions, when painting `object-fit` objects into `contain:size`d elements. I think we need some more spec language somewhere here, although admittedly my understanding of the precise meaning of [concrete object size](https://www.w3.org/TR/css-images-3/#concrete-object-size) is provisional at best.

**Comment by mirisuzanne** [comment]
The reason for `object-fit` and related properties is to help align and size the image content inside a box that has different dimensions, or a different aspect ratio - because of some other specified styles. When we start applying those box layout dimensions to the image content, we lose the purpose of the feature.

**Comment by tabatkins** [comment]
Agreed, 'object-fit' 'itself clearly requires working on the actual image dimensions, or else it's worthless. The c-i-s dimensions are just to force layout stability. They only have an effect on the image's dimensions as a side-effect, since images default to stretching into their layout dimensions.

I'd have to poke around to figure out what, if anything, needs fixing in the specs; the language around intrinsic/natural dimensions is a bit wonky and there probably needs to be more "layers" of terminology here.

**Comment by Loirooriol** [comment]
Some previous discussion in https://github.com/w3c/csswg-drafts/issues/6257

**Comment by mirisuzanne** [comment]
We could close one of the issues, and add agenda+ to the other? It seems like there's consensus this is broken. And it seems like it should be clarified _somewhere_.

**Comment by vmpstr** [comment]
Is this a consequence of `contain-intrinsic-size` or size containment? [Size containment](https://www.w3.org/TR/css-contain-2/#containment-size) says that 

> [Replaced elements](https://www.w3.org/TR/css-display-3/#replaced-element) must be treated as having an [natural](https://www.w3.org/TR/css-images-3/#natural-dimensions) width and height of 0 and no [natural aspect ratio](https://www.w3.org/TR/css-images-3/#natural-aspect-ratio).

[contain-intrinsic-size definition](https://drafts.csswg.org/css-sizing-4/#explicit-intrinsic-inner-size) says:

> These properties allow elements with [size containment](https://drafts.csswg.org/css-contain-2/#size-containment) to specify an explicit intrinsic inner size, causing the box to size as if its in-flow content totals to a width and height matching the specified [explicit intrinsic inner size](https://drafts.csswg.org/css-sizing-4/#explicit-intrinsic-inner-size) (rather than sizing as if it were empty).

I couldn't find the reference for contain-intrinsic-size and natural size interactions.

In particular, in the codepen example (on Chromium), I can't get a different effect if I change contain-intrinsic-size to something like 800px 400px. It always remains a square.

**Comment by eeeps** [comment]
@vmpstr Try refreshing? This has been flaky for me [before](https://github.com/whatwg/html/issues/10079#issuecomment-1901148238). Not sure if that's CodePen's fault, or Chromiums. Anyways here's proof that it does sometimes have an effect: https://o.img.rodeo/c2axgu9ms1l3ojujbjjn.png https://codepen.io/eeeps/pen/abxWVmN?editors=1101

**Comment by Loirooriol** [comment]
> It always remains a square

Open devtools and set `display: none`, then unset it.

> Is this a consequence of `contain-intrinsic-size` or size containment?

Size containment, which may be affected by `contain-intrinsic-size`.

> I couldn't find the reference for contain-intrinsic-size and natural size interactions.

See the resolution in https://github.com/w3c/csswg-drafts/issues/7519#issuecomment-1202721367.

But as I argued in https://github.com/w3c/csswg-drafts/issues/6257#issuecomment-850709682, this is just for sizing the `<img>`. Once we know it's size, the spec says to lay out normally. So `object-fit` should work. And since the example uses `width` and `height`, `contain-intrinsic-size` should have no effect at all.

**Comment by vmpstr** [comment]
Got it working, thanks all!

And yes, it makes sense that `contain-intrinsic-size` affects the img element size not the interaction with `object-fit`

**Comment by mirisuzanne** [comment]
I think the desired resolution here would be what @bfgeek proposed in https://github.com/w3c/csswg-drafts/issues/6257#issuecomment-849150205 - though it's not clear to me which spec (if any) needs to be clarified. Maybe we're just confirming this, and adding the WPT to prove it? :)

**Comment by eeeps** [comment]
Ran into this today in Safari and submitted https://bugs.webkit.org/show_bug.cgi?id=276681

**Comment by eeeps** [comment]
A patch was submitted to the WebKit bug today. https://bugs.webkit.org/show_bug.cgi?id=276681#c2

So perhaps soon we will have three implementations, tests, but no spec yet.

**Comment by Loirooriol** [comment]
Even though clarifying the spec would be good, it already implies the resolved behavior, as I explained in https://github.com/w3c/csswg-drafts/issues/6257#issuecomment-850709682
